### ISSUE TITLE
concurrent processing issue

### ISSUE THREAD

**Issue by orelant** [dailyburn/ratchet#37]
Hi there,

Im trying to run FuncTransformer concurrently and can't get it working.
Could someone point me out what I am missing, please.

```go
	logger.LogLevel = logger.LevelDebug

	db1, db2 := database.SetupDatabases()
	db2.Query("delete from users")

	read := processors.NewSQLReader(db1, "SELECT id, name FROM users")
	read.BatchSize = 1
	transform := processors.NewFuncTransformer(func(d data.JSON) data.JSON {
		fmt.Println(string(d))
		json := data.JSON(string(d))
		time.Sleep(3 * time.Second)
		return json
	})
	transform.ConcurrencyLevel = 2

	write := processors.NewSQLWriter(db2, "users")

	layout, err := ratchet.NewPipelineLayout(
		ratchet.NewPipelineStage(
			ratchet.Do(read).Outputs(transform),
		),
		ratchet.NewPipelineStage(
			ratchet.Do(transform).Outputs(write),
		),
		ratchet.NewPipelineStage(
			ratchet.Do(write),
		),
	)
	if err != nil {
		// layout is invalid
		panic(err.Error())
	}

	// Finally, create and run the Pipeline
	pipeline := ratchet.NewBranchingPipeline(layout)
	err = <-pipeline.Run()

```

I expect that code above will be executed within 3 secs but it takes 6 seconds.
There is a debug output:

```
2017/04/21 15:48:59 [Pipeline : connecting stages]
2017/04/21 15:48:59 [Pipeline : sending GO to SQLReader]
2017/04/21 15:48:59 [Pipeline - stage 3 SQLWriter waiting to receive data]
2017/04/21 15:48:59 [Pipeline - stage 1 SQLReader waiting to receive data]
2017/04/21 15:48:59 [Pipeline - stage 1 SQLReader received data]
2017/04/21 15:48:59 [dataProcessor: processData SQLReader with concurrency = 0]
2017/04/21 15:48:59 [SQLReader: Running -  SELECT id, name FROM users]
2017/04/21 15:48:59 [Pipeline - stage 2 FuncTransformer waiting to receive data]
2017/04/21 15:48:59 [Pipeline - stage 1 SQLReader input closed, calling Finish]
2017/04/21 15:48:59 [Pipeline - stage 1 SQLReader closing output]
2017/04/21 15:48:59 [Pipeline - stage 2 FuncTransformer received data]
2017/04/21 15:48:59 [dataProcessor: processData FuncTransformer with concurrency = 2]
2017/04/21 15:48:59 [dataProcessor: processData FuncTransformer waiting for work]
2017/04/21 15:48:59 [dataProcessor: processData FuncTransformer work obtained]
[{"id":1,"name":"Aaron"}]
2017/04/21 15:48:59 [dataProcessor: processData FuncTransformer waiting to receive data on result chan]
2017/04/21 15:49:02 [dataProcessor: processData FuncTransformer received data on result chan]
2017/04/21 15:49:02 [dataProcessor: processData FuncTransformer done, releasing work]
2017/04/21 15:49:02 [dataProcessor: sendResults checking for valid data to send]
2017/04/21 15:49:02 [dataHandler: sendResults sending data]
2017/04/21 15:49:02 [Pipeline - stage 2 FuncTransformer received data]
2017/04/21 15:49:02 [dataProcessor: processData FuncTransformer with concurrency = 2]
2017/04/21 15:49:02 [dataProcessor: processData FuncTransformer waiting for work]
2017/04/21 15:49:02 [dataProcessor: processData FuncTransformer work obtained]
[{"id":2,"name":"Bob"}]
2017/04/21 15:49:02 [Pipeline - stage 3 SQLWriter received data]
2017/04/21 15:49:02 [dataProcessor: processData SQLWriter with concurrency = 0]
2017/04/21 15:49:02 [SQLWriter: Writing data...]
2017/04/21 15:49:02 [SQLWriter: normal data scenario]
2017/04/21 15:49:02 [dataProcessor: processData FuncTransformer waiting to receive data on result chan]
2017/04/21 15:49:02 [SQLInsertData: building INSERT for len(objects) = 1]
2017/04/21 15:49:02 [SQLInsertData: INSERT INTO users(id,name) VALUES(?,?) ON DUPLICATE KEY UPDATE `id`=VALUES(`id`),`name`=VALUES(`name`)]
2017/04/21 15:49:02 [SQLInsertData: values [1 Aaron]]
2017/04/21 15:49:02 [SQLInsertData: rows affected = 1, last insert ID = 0]
2017/04/21 15:49:02 [SQLWriter: Write complete]
2017/04/21 15:49:05 [dataProcessor: processData FuncTransformer received data on result chan]
2017/04/21 15:49:05 [dataProcessor: processData FuncTransformer done, releasing work]
2017/04/21 15:49:05 [dataProcessor: sendResults checking for valid data to send]
2017/04/21 15:49:05 [dataHandler: sendResults sending data]
2017/04/21 15:49:05 [Pipeline - stage 2 FuncTransformer input closed, calling Finish]
2017/04/21 15:49:05 [Pipeline - stage 2 FuncTransformer closing output]
2017/04/21 15:49:05 [Pipeline - stage 3 SQLWriter received data]
2017/04/21 15:49:05 [dataProcessor: processData SQLWriter with concurrency = 0]
2017/04/21 15:49:05 [SQLWriter: Writing data...]
2017/04/21 15:49:05 [SQLWriter: normal data scenario]
2017/04/21 15:49:05 [SQLInsertData: building INSERT for len(objects) = 1]
2017/04/21 15:49:05 [SQLInsertData: INSERT INTO users(id,name) VALUES(?,?) ON DUPLICATE KEY UPDATE `id`=VALUES(`id`),`name`=VALUES(`name`)]
2017/04/21 15:49:05 [SQLInsertData: values [2 Bob]]
2017/04/21 15:49:05 [SQLInsertData: rows affected = 1, last insert ID = 0]
2017/04/21 15:49:05 [SQLWriter: Write complete]
2017/04/21 15:49:05 [Pipeline - stage 3 SQLWriter input closed, calling Finish]
2017/04/21 15:49:05 [Pipeline - stage 3 SQLWriter closing output]
```

Thank you in advance.

**Comment by orelant** [comment]
@rkulla @sturgill guys could you help me to understand that, please?

**Comment by sturgill** [comment]
Hey @orelant, we'll try to take a look this weekend / early next week. We have to get some stuff out the door by the end of month.

Thanks for submitting the issue. We'll update here when more is known.

**Comment by llonchj** [comment]
+1

**Comment by orelant** [comment]
@rkulla @sturgill hey guys, did you have a chance to look at it? Thanks!

**Comment by rkulla** [comment]
Sorry, I finally just got a chance to look at it.  I think you just need to set `read.BatchSize` to 2 instead of 1.

**Comment by orelant** [comment]
thanks for the response @rkulla. Thus, if I set batch size to 2 FuncTransformer will be executed once with two records. Here is the output:

```
2017/05/12 13:34:27 [Pipeline - stage 3 SQLWriter waiting to receive data]
2017/05/12 13:34:27 [Pipeline - stage 1 SQLReader waiting to receive data]
2017/05/12 13:34:27 [Pipeline - stage 2 FuncTransformer waiting to receive data]
2017/05/12 13:34:27 [Pipeline - stage 1 SQLReader received data]
2017/05/12 13:34:27 [dataProcessor: processData SQLReader with concurrency = 0]
2017/05/12 13:34:27 [SQLReader: Running -  SELECT id, name FROM users]
2017/05/12 13:34:27 [Pipeline - stage 1 SQLReader input closed, calling Finish]
2017/05/12 13:34:27 [Pipeline - stage 1 SQLReader closing output]
2017/05/12 13:34:27 [Pipeline - stage 2 FuncTransformer received data]
2017/05/12 13:34:27 [dataProcessor: processData FuncTransformer with concurrency = 2]
2017/05/12 13:34:27 [dataProcessor: processData FuncTransformer waiting for work]
2017/05/12 13:34:27 [dataProcessor: processData FuncTransformer work obtained]
[{"id":2,"name":"Aaron"},{"id":3,"name":"Bob"}]
2017/05/12 13:34:27 [dataProcessor: processData FuncTransformer waiting to receive data on result chan]
2017/05/12 13:34:30 [dataProcessor: processData FuncTransformer received data on result chan]
2017/05/12 13:34:30 [dataProcessor: processData FuncTransformer done, releasing work]
2017/05/12 13:34:30 [dataProcessor: sendResults checking for valid data to send]
2017/05/12 13:34:30 [dataHandler: sendResults sending data]
2017/05/12 13:34:30 [Pipeline - stage 2 FuncTransformer input closed, calling Finish]
2017/05/12 13:34:30 [Pipeline - stage 2 FuncTransformer closing output]
2017/05/12 13:34:30 [Pipeline - stage 3 SQLWriter received data]
2017/05/12 13:34:30 [dataProcessor: processData SQLWriter with concurrency = 0]
2017/05/12 13:34:30 [SQLWriter: Writing data...]
2017/05/12 13:34:30 [SQLWriter: normal data scenario]
2017/05/12 13:34:30 [SQLInsertData: building INSERT for len(objects) = 2]
2017/05/12 13:34:30 [SQLInsertData: INSERT INTO users(id,name) VALUES(?,?),(?,?) ON DUPLICATE KEY UPDATE `id`=VALUES(`id`),`name`=VALUES(`name`)]
2017/05/12 13:34:30 [SQLInsertData: values [2 Aaron 3 Bob]]
2017/05/12 13:34:30 [SQLInsertData: rows affected = 2, last insert ID = 3]
2017/05/12 13:34:30 [SQLWriter: Write complete]
2017/05/12 13:34:30 [Pipeline - stage 3 SQLWriter input closed, calling Finish]
2017/05/12 13:34:30 [Pipeline - stage 3 SQLWriter closing output]
```

Still dont get it how is the concurrency  supposed to work... Any thoughts?

**Comment by rkulla** [comment]
From what I understand, when you set `transform.ConcurrencyLevel = 2`, you made it so the pipeline allows the transform stage to execute up to 2 ProcessData() calls concurrently. Which is why you see "dataProcessor: processData FuncTransformer waiting to receive data on result chan".  Your other data processors show concurrency = 0 because you didn't specify a concurrency level for those stages.

So ratchet will process the data in a concurrent queue of goroutines. Whereas if you had set it to 1 (or less), then you're saying no concurrency is needed, simply call ProcessData and return.   This happens because the data processors embed the `ConcurrentDataProcessor` functionality. See: https://github.com/dailyburn/ratchet/blob/8b15610f137081b995bcfbc9b4e5885cf080c196/concurrent_data_processor.go#L11

But still remember that "Each data processor is receiving, processing, and then sending data to the next stage in the pipeline. All data processors are running in their own goroutine, so all processing is happening concurrently. Go channels are connecting each stage of processing"

**Comment by orelant** [comment]
I understand what you are saying and this is exactly how I expect it should work. Nevertheless, that means the example at top is supposed to take 3 seconds instead of 6. I tried to set ConcurrencyLevel to 2 for either both 1 and 3 stage, no luck. No matter what I do FuncTransfomer with ConcurrencyLevel greater than 1 always executes successively. That would be great if you could provide a working example. That would be a good addition to the readme. Thank you for your time!

**Comment by llonchj** [comment]
+1

**Comment by rkulla** [comment]
Ok, I will have to go through it when I can and get back to you. I haven't used it and mostly just use custom transformers and set batchSize on things.

**Comment by sturgill** [comment]
Thanks for the PR, @jimberlage. This is merged into the develop branch (and release 3.0.0 which should be cut soon).

**Comment by jimberlage** [comment]
No prob.  Thanks for the library + the review!
